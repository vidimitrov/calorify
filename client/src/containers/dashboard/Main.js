/* eslint react/no-array-index-key: 0 */

import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Redirect } from '@reach/router';
import moment from 'moment';
import _ from 'lodash';
import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import IconButton from '@material-ui/core/IconButton';
import MenuItem from '@material-ui/core/MenuItem';
import Menu from '@material-ui/core/Menu';
import AccountCircle from '@material-ui/icons/AccountCircle';
import MoreVert from '@material-ui/icons/MoreVert';
import AddIcon from '@material-ui/icons/Add';
import Logo from '../../components/dashboard/Logo';
import Wrapper from '../../components/dashboard/Wrapper';
import CustomTypography from '../../components/dashboard/CustomTypography';
import FabButton from '../../components/dashboard/FabButton';
import MealsList from '../../components/dashboard/MealsList';
import ScrollContainer from '../../components/dashboard/ScrollContainer';
import GroupHeading from '../../components/dashboard/GroupHeading';
import Card from '../../components/dashboard/Card/Card';
import CardInfo from '../../components/dashboard/Card/CardInfo';
import CardDate from '../../components/dashboard/Card/CardDate';
import CardActions from '../../components/dashboard/Card/CardActions';
import CardStatus from '../../components/dashboard/Card/CardStatus';
import logo from '../../assets/img/logo.png';
import { logout as logoutActionCreator } from '../../actions/authentication';
import {
  fetchMeals as fetchMealsActionCreator,
  removeMeal as removeMealActionCreator,
} from '../../actions/meals';

export class Main extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      anchorEl: null,
    };

    this.handleMenu = this.handleMenu.bind(this);
    this.handleClose = this.handleClose.bind(this);
    this.isInRange = this.isInRange.bind(this);
  }

  async componentDidMount() {
    const { getAllMeals } = this.props;
    try {
      await getAllMeals();
    } catch (error) {
      // TODO: Show error snackbar here
      throw error;
    }
  }

  handleMenu(event) {
    this.setState({
      anchorEl: event.currentTarget,
    });
  }

  handleClose() {
    this.setState({
      anchorEl: null,
    });
  }

  isInRange(date) {
    const { meals, user } = this.props;
    const filteredMeals = meals.filter(m => m.date === date);
    const totalCalories = filteredMeals
      .reduce((accumulator, currentValue) => accumulator + currentValue.number_of_calories, 0);

    return totalCalories < user.daily_calories_limit;
  }

  render() {
    const {
      navigate,
      user,
      meals,
      logout,
      removeMeal,
    } = this.props;
    const { anchorEl } = this.state;
    const open = Boolean(anchorEl);
    const mealsGroupedByDate = _.groupBy(meals, 'date');
    const groupDates = Object.keys(mealsGroupedByDate).sort((a, b) => moment(b).diff(moment(a)));

    if (!user) {
      return (
        <Redirect to="/auth/login" />
      );
    }

    return (
      <Wrapper>
        <AppBar position="static">
          <Toolbar>
            <Logo src={logo} alt="logo" />
            <CustomTypography variant="h6" color="secondary">
              Calorify
            </CustomTypography>
            <div>
              <IconButton
                aria-owns={open ? 'menu-appbar' : undefined}
                aria-haspopup="true"
                onClick={() => navigate('/settings')}
                color="secondary"
              >
                <AccountCircle />
              </IconButton>
              <IconButton
                aria-owns={open ? 'menu-appbar' : undefined}
                aria-haspopup="true"
                onClick={this.handleMenu}
                color="secondary"
              >
                <MoreVert />
              </IconButton>
              <Menu
                id="menu-appbar"
                anchorEl={anchorEl}
                anchorOrigin={{
                  vertical: 'top',
                  horizontal: 'right',
                }}
                transformOrigin={{
                  vertical: 'top',
                  horizontal: 'right',
                }}
                open={open}
                onClose={this.handleClose}
              >
                <MenuItem onClick={() => logout().then(() => navigate('/auth/login'))}>Logout</MenuItem>
              </Menu>
            </div>
          </Toolbar>
        </AppBar>
        <MealsList>
          <ScrollContainer>
            {groupDates.map((gDate, index) => (
              <div key={index}>
                <GroupHeading>{(moment(gDate).calendar().split(' at'))[0]}</GroupHeading>
                {mealsGroupedByDate[gDate].map(meal => (
                  <Card key={meal.id}>
                    <CardStatus inRange={this.isInRange(meal.date)} />
                    <CardInfo text={meal.text} calories={meal.number_of_calories} />
                    <CardDate date={meal.updated_at} />
                    <CardActions
                      onEditHandler={() => navigate(`/meals/${meal.id}`)}
                      onDeleteHandler={() => removeMeal(meal.id).then(() => this.forceUpdate())}
                    />
                  </Card>
                ))}
              </div>
            ))}
          </ScrollContainer>
        </MealsList>
        <FabButton
          variant="fab"
          color="primary"
          aria-label="Add"
          onClick={() => {
            navigate('/create-meal');
          }}
        >
          <AddIcon />
        </FabButton>
      </Wrapper>
    );
  }
}

Main.propTypes = {
  user: PropTypes.shape({
    id: PropTypes.string.isRequired,
    role: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    email: PropTypes.string.isRequired,
    daily_calories_limit: PropTypes.number.isRequired,
  }),
  meals: PropTypes.arrayOf(PropTypes.shape({
    id: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired,
    date: PropTypes.string.isRequired,
    time: PropTypes.string.isRequired,
    number_of_calories: PropTypes.number.isRequired,
  })),
  navigate: PropTypes.func.isRequired,
  logout: PropTypes.func.isRequired,
  getAllMeals: PropTypes.func.isRequired,
  removeMeal: PropTypes.func.isRequired,
};

Main.defaultProps = {
  user: null,
  meals: [],
};

const mapStateToProps = state => ({
  user: state.user.data,
  meals: state.meals.data,
});

const mapDispatchToProps = dispatch => ({
  logout: () => dispatch(logoutActionCreator()),
  getAllMeals: () => dispatch(fetchMealsActionCreator()),
  removeMeal: mealId => dispatch(removeMealActionCreator(mealId)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Main);
