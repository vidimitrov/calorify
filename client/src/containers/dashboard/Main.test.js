import React from 'react';
import renderer from 'react-test-renderer';
import { Main } from './Main';

it('Main container renders correctly', () => {
  const props = {
    logout: () => { },
    getAllMeals: () => { },
    removeMeal: () => { },
    filterMeals: () => { },
    navigate: () => { },
    account: {
      id: 'some-fake-id',
      name: 'John',
      dailyCaloriesLimit: 2000,
      role: 'user',
      email: 'john@email.com',
      createdAt: '2018-09-10',
    },
    meals: [{
      id: 'some-fake-meal-id',
      name: 'some-name',
      date: '2018-05-10',
      time: '13:35',
      calories: 320,
    }],
    initialDate: '2018-09-10',
  };
  const tree = renderer.create(
    <Main {...props} />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
