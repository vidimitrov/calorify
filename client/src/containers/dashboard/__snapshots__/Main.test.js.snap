// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main container renders correctly 1`] = `
<div
  className="Wrapper-m60uq5-0 hkWJUG"
  data-testid="dashboard"
>
  <header
    className="MuiPaper-root-10 MuiPaper-elevation4-16 MuiAppBar-root-1 MuiAppBar-positionStatic-5 MuiAppBar-colorPrimary-8"
  >
    <div
      className="MuiToolbar-root-37 MuiToolbar-regular-39 MuiToolbar-gutters-38"
    >
      <img
        alt="logo"
        className="Logo-sc-1vcm3wd-0 lmrySJ"
        src="logo.png"
      />
      <h6
        className="MuiTypography-root-41 MuiTypography-h6-58 MuiTypography-colorSecondary-72 CustomTypography-e31nmz-0 kgGAHm"
      >
        Calorify
      </h6>
      <div>
        <button
          className="MuiButtonBase-root-82 MuiIconButton-root-76 MuiIconButton-colorSecondary-79"
          disabled={false}
          name="account-settings"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiIconButton-label-81"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-85"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 3c1.66 0 3 1.34 3 3s-1.34 3-3 3-3-1.34-3-3 1.34-3 3-3zm0 14.2c-2.5 0-4.71-1.28-6-3.22.03-1.99 4-3.08 6-3.08 1.99 0 5.97 1.09 6 3.08-1.29 1.94-3.5 3.22-6 3.22z"
              />
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-179"
          />
        </button>
        <button
          className="MuiButtonBase-root-82 MuiIconButton-root-76 MuiIconButton-colorSecondary-79"
          disabled={false}
          name="logout-button"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiIconButton-label-81"
          >
            <svg
              aria-hidden="true"
              className="MuiSvgIcon-root-85"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M0 0h24v24H0z"
                fill="none"
              />
              <path
                d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
              />
            </svg>
          </span>
          <span
            className="MuiTouchRipple-root-179"
          />
        </button>
      </div>
    </div>
  </header>
  <div
    className="Filters-sc-14xsfyg-0 dkYwmc"
  >
    <div
      className="FiltersControls-sc-1s7sukz-0 ktuGWe"
    >
      <div
        className="MuiFormControl-root-94"
      >
        <label
          className="MuiFormLabel-root-109 MuiFormLabel-filled-113 MuiInputLabel-root-98 MuiInputLabel-formControl-103 MuiInputLabel-animated-106 MuiInputLabel-shrink-105"
          data-shrink={true}
        >
          From
        </label>
        <div
          className="MuiInputBase-root-129 MuiInput-root-116 MuiInput-underline-120 MuiInputBase-formControl-130 MuiInput-formControl-117"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input-139 MuiInput-input-124 MuiInputBase-inputType-142 MuiInput-inputType-127"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="date"
            value="2018-11-11"
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root-94"
      >
        <label
          className="MuiFormLabel-root-109 MuiFormLabel-filled-113 MuiInputLabel-root-98 MuiInputLabel-formControl-103 MuiInputLabel-animated-106 MuiInputLabel-shrink-105"
          data-shrink={true}
        >
          To
        </label>
        <div
          className="MuiInputBase-root-129 MuiInput-root-116 MuiInput-underline-120 MuiInputBase-formControl-130 MuiInput-formControl-117"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input-139 MuiInput-input-124 MuiInputBase-inputType-142 MuiInput-inputType-127"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="date"
            value="2018-11-11"
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root-94"
      >
        <label
          className="MuiFormLabel-root-109 MuiFormLabel-filled-113 MuiInputLabel-root-98 MuiInputLabel-formControl-103 MuiInputLabel-animated-106 MuiInputLabel-shrink-105"
          data-shrink={true}
        >
          From
        </label>
        <div
          className="MuiInputBase-root-129 MuiInput-root-116 MuiInput-underline-120 MuiInputBase-formControl-130 MuiInput-formControl-117"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input-139 MuiInput-input-124 MuiInputBase-inputType-142 MuiInput-inputType-127"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="time"
            value="00:00"
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root-94"
      >
        <label
          className="MuiFormLabel-root-109 MuiFormLabel-filled-113 MuiInputLabel-root-98 MuiInputLabel-formControl-103 MuiInputLabel-animated-106 MuiInputLabel-shrink-105"
          data-shrink={true}
        >
          To
        </label>
        <div
          className="MuiInputBase-root-129 MuiInput-root-116 MuiInput-underline-120 MuiInputBase-formControl-130 MuiInput-formControl-117"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            className="MuiInputBase-input-139 MuiInput-input-124 MuiInputBase-inputType-142 MuiInput-inputType-127"
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="time"
            value="23:59"
          />
        </div>
      </div>
    </div>
    <div>
      <button
        aria-haspopup="true"
        className="MuiButtonBase-root-82 MuiIconButton-root-76"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-81"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-85"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 18h4v-2h-4v2zM3 6v2h18V6H3zm3 7h12v-2H6v2z"
            />
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-179"
        />
      </button>
    </div>
  </div>
  <div
    className="MealsList-hfcb7x-0 eOsDlS"
  >
    <div
      className="ScrollContainer__Form-sc-1eluxdx-0 leIDbD"
    >
      <div>
        <h2
          className="GroupHeading__InputWrapper-sc-1140iy9-0 bAAVPR"
        >
          05/10/2018
        </h2>
        <div
          className="Card-sc-1astb0w-0 bRrYbQ"
          name="meal-card"
        >
          <div
            className="CardStatus-sc-1qy5vwr-0 AykxL"
          />
          <div
            className="CardInfo__Wrapper-sc-1xghcvm-0 oWCwm"
          >
            <h1
              className="CardInfo__Calories-sc-1xghcvm-2 jiDjuh"
              name="meal-calories"
            >
              320
               
              kCal
            </h1>
            <h3
              className="CardInfo__Name-sc-1xghcvm-1 jsbjuo"
              name="meal-name"
            >
              some-name
            </h3>
          </div>
          <div
            className="CardDate__Wrapper-pveplr-0 gvOOJP"
          >
            <h3
              className="CardDate__FormattedDate-pveplr-1 kxClMg"
            >
              05/10/2018
            </h3>
          </div>
          <div
            className="CardActions__Wrapper-il8ivv-0 iSeIUb"
          >
            <button
              className="MuiButtonBase-root-82 MuiIconButton-root-76"
              disabled={false}
              name="edit-meal"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex="0"
              type="button"
            >
              <span
                className="MuiIconButton-label-81"
              >
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root-85"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                  />
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                </svg>
              </span>
              <span
                className="MuiTouchRipple-root-179"
              />
            </button>
            <button
              className="MuiButtonBase-root-82 MuiIconButton-root-76"
              disabled={false}
              name="delete-meal"
              onBlur={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              tabIndex="0"
              type="button"
            >
              <span
                className="MuiIconButton-label-81"
              >
                <svg
                  aria-hidden="true"
                  className="MuiSvgIcon-root-85"
                  focusable="false"
                  role="presentation"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                  />
                  <path
                    d="M0 0h24v24H0z"
                    fill="none"
                  />
                </svg>
              </span>
              <span
                className="MuiTouchRipple-root-179"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <button
    aria-label="Add"
    className="MuiButtonBase-root-82 MuiButton-root-146 MuiButton-fab-163 MuiButton-contained-157 MuiButton-containedPrimary-158 MuiButton-raised-160 MuiButton-raisedPrimary-161 FabButton-sc-1kemqdv-0 jPpghY"
    disabled={false}
    name="add-meal-fab"
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-147"
    >
      <svg
        aria-hidden="true"
        className="MuiSvgIcon-root-85"
        focusable="false"
        role="presentation"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
        <path
          d="M0 0h24v24H0z"
          fill="none"
        />
      </svg>
    </span>
    <span
      className="MuiTouchRipple-root-179"
    />
  </button>
</div>
`;
